File: sdks/typescript/test/signature.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { verifyWebhookSignature } from "../src/signature.js";

interface SignatureVector {
  name: string;
  secret: string;
  timestamp: string;
  body: string;
  signature_hex: string;
  header_value: string;
}

function loadSignatureVectors(): SignatureVector[] {
  const vectorsPath = fileURLToPath(new URL("../../testdata/signature_vectors.json", import.meta.url));
  const vectors = JSON.parse(readFileSync(vectorsPath, "utf8")) as SignatureVector[];
  if (vectors.length === 0) {
    throw new Error("signature_vectors.json is empty");
  }
  return vectors;
}

describe("verifyWebhookSignature against shared cross-language vectors", () => {
  const vectors = loadSignatureVectors();

  it.each(vectors.map((v) => [v.name, v] as const))("%s", (_name, vector) => {
    const isValid = verifyWebhookSignature(
      vector.secret,
      vector.timestamp,
      vector.body,
      vector.header_value,
    );
    expect(isValid).toBe(true);
  });
});

describe("verifyWebhookSignature rejects invalid input", () => {
  const [vector] = loadSignatureVectors();

  it("rejects a tampered body", () => {
    const isValid = verifyWebhookSignature(
      vector.secret,
      vector.timestamp,
      vector.body + "tampered",
      vector.header_value,
    );
    expect(isValid).toBe(false);
  });

  it("rejects the wrong secret", () => {
    const isValid = verifyWebhookSignature(
      "wrong-secret",
      vector.timestamp,
      vector.body,
      vector.header_value,
    );
    expect(isValid).toBe(false);
  });

  it.each([
    ["missing sha256= prefix", vector.signature_hex],
    ["wrong algorithm prefix", `sha1=${vector.signature_hex}`],
    ["undecodable hex", "sha256=not-hex!!"],
    ["empty header", ""],
  ])("rejects a malformed header (%s)", (_label, malformedHeader) => {
    const isValid = verifyWebhookSignature(vector.secret, vector.timestamp, vector.body, malformedHeader);
    expect(isValid).toBe(false);
  });
});

File: sdks/go/signature_test.go
package notifyd

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// signatureVector mirrors one entry in sdks/testdata/signature_vectors.json.
type signatureVector struct {
	Name         string `json:"name"`
	Secret       string `json:"secret"`
	Timestamp    string `json:"timestamp"`
	Body         string `json:"body"`
	SignatureHex string `json:"signature_hex"`
	HeaderValue  string `json:"header_value"`
}

func loadSignatureVectors(t *testing.T) []signatureVector {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "testdata", "signature_vectors.json"))
	if err != nil {
		t.Fatalf("reading shared signature vectors: %v", err)
	}
	var vectors []signatureVector
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("parsing shared signature vectors: %v", err)
	}
	if len(vectors) == 0 {
		t.Fatal("signature_vectors.json is empty")
	}
	return vectors
}

func TestVerifyWebhookSignature_SharedVectors(t *testing.T) {
	for _, v := range loadSignatureVectors(t) {
		t.Run(v.Name, func(t *testing.T) {
			err := VerifyWebhookSignature(v.Secret, v.Timestamp, []byte(v.Body), v.HeaderValue)
			if err != nil {
				t.Fatalf("VerifyWebhookSignature: %v (expected valid signature %s)", err, v.HeaderValue)
			}
		})
	}
}

func TestVerifyWebhookSignature_RejectsTamperedBody(t *testing.T) {
	vectors := loadSignatureVectors(t)
	v := vectors[0]

	err := VerifyWebhookSignature(v.Secret, v.Timestamp, []byte(v.Body+"tampered"), v.HeaderValue)
	if err != ErrInvalidSignature {
		t.Fatalf("got %v, want ErrInvalidSignature", err)
	}
}

func TestVerifyWebhookSignature_RejectsWrongSecret(t *testing.T) {
	vectors := loadSignatureVectors(t)
	v := vectors[0]

	err := VerifyWebhookSignature("wrong-secret", v.Timestamp, []byte(v.Body), v.HeaderValue)
	if err != ErrInvalidSignature {
		t.Fatalf("got %v, want ErrInvalidSignature", err)
	}
}

func TestVerifyWebhookSignature_RejectsMalformedHeader(t *testing.T) {
	vectors := loadSignatureVectors(t)
	v := vectors[0]

	for _, malformed := range []string{
		v.SignatureHex,           // missing "sha256=" prefix
		"sha1=" + v.SignatureHex, // wrong algorithm prefix
		"sha256=not-hex!!",       // undecodable hex
		"",
	} {
		err := VerifyWebhookSignature(v.Secret, v.Timestamp, []byte(v.Body), malformed)
		if err != ErrInvalidSignature {
			t.Errorf("header %q: got %v, want ErrInvalidSignature", malformed, err)
		}
	}
}
